File: orion-ui/packages/orion-design/src/models/FlowRun.ts
import { IState } from '@/models/State'
import { StateType } from '@/models/StateType'

export interface IFlowRun {
  id: string,
  flowId: string,
  deploymentId: string | null,
  flowVersion: string | null,
  idempotencyKey: string | null,
  expectedStartTime: string | null,
  nextScheduledStartTime: string | null,
  parameters: unknown,
  autoScheduled: boolean | null,
  context: unknown,
  empiricalConfig: unknown,
  empiricalPolicy: unknown,
  estimatedRunTime: number | null,
  estimatedStartTimeDelta: number | null,
  totalRunTime: number | null,
  startTime: Date | null,
  endTime: Date | null,
  name: string | null,
  parentTaskRunId: string | null,
  stateId: string | null,
  stateType: StateType | null,
  state: IState | null,
  tags: string[] | null,
  runCount: number | null,
  created: Date,
  updated: Date,
}

export class FlowRun implements IFlowRun {
  public readonly id: string
  public readonly flowId: string
  public readonly deploymentId: string | null
  public flowVersion: string | null
  public idempotencyKey: string | null
  public expectedStartTime: string | null
  public nextScheduledStartTime: string | null
  public parameters: unknown
  public autoScheduled: boolean | null
  public context: unknown
  public empiricalConfig: unknown
  public empiricalPolicy: unknown
  public estimatedRunTime: number | null
  public estimatedStartTimeDelta: number | null
  public totalRunTime: number | null
  public startTime: Date | null
  public endTime: Date | null
  public name: string | null
  public parentTaskRunId: string | null
  public stateId: string | null
  public stateType: StateType | null
  public state: IState | null
  public tags: string[] | null
  public runCount: number | null
  public created: Date
  public updated: Date

  public constructor(flow: IFlowRun) {
    this.id = flow.id
    this.deploymentId = flow.deploymentId
    this.flowId = flow.flowId
    this.flowVersion = flow.flowVersion
    this.idempotencyKey = flow.idempotencyKey
    this.expectedStartTime = flow.expectedStartTime
    this.nextScheduledStartTime = flow.nextScheduledStartTime
    this.parameters = flow.parameters
    this.autoScheduled = flow.autoScheduled
    this.context = flow.context
    this.empiricalConfig = flow.empiricalConfig
    this.empiricalPolicy = flow.empiricalPolicy
    this.estimatedRunTime = flow.estimatedRunTime
    this.estimatedStartTimeDelta = flow.estimatedStartTimeDelta
    this.totalRunTime = flow.totalRunTime
    this.startTime = flow.startTime
    this.endTime = flow.endTime
    this.name = flow.name
    this.parentTaskRunId = flow.parentTaskRunId
    this.stateId = flow.stateId
    this.stateType = flow.stateType
    this.state = flow.state
    this.tags = flow.tags
    this.runCount = flow.runCount
    this.created = flow.created
    this.updated = flow.updated
  }

  public get duration(): number {
    return this.totalRunTime ?? this.estimatedRunTime ?? 0
  }
}

File: src/prefect/client/schemas/objects/flow_run.py
import datetime
from typing import (
    Any,
    Dict,
    List,
    Optional,
)
from uuid import UUID

from pydantic import (
    Field,
    field_validator,
)
from pydantic_extra_types.pendulum_dt import DateTime

from prefect._internal.schemas.bases import ObjectBaseModel
from prefect._internal.schemas.fields import CreatedBy
from prefect._internal.schemas.validators import (
    get_or_create_run_name,
)
from prefect.utilities.names import generate_slug

from .flow_run_policy import FlowRunPolicy
from .state import State
from .state_type import StateType


class FlowRun(ObjectBaseModel):
    name: str = Field(
        default_factory=lambda: generate_slug(2),
        description=(
            "The name of the flow run. Defaults to a random slug if not specified."
        ),
        examples=["my-flow-run"],
    )
    flow_id: UUID = Field(default=..., description="The id of the flow being run.")
    state_id: Optional[UUID] = Field(
        default=None, description="The id of the flow run's current state."
    )
    deployment_id: Optional[UUID] = Field(
        default=None,
        description=(
            "The id of the deployment associated with this flow run, if available."
        ),
    )
    deployment_version: Optional[str] = Field(
        default=None,
        description="The version of the deployment associated with this flow run.",
        examples=["1.0"],
    )
    work_queue_name: Optional[str] = Field(
        default=None, description="The work queue that handled this flow run."
    )
    flow_version: Optional[str] = Field(
        default=None,
        description="The version of the flow executed in this flow run.",
        examples=["1.0"],
    )
    parameters: Dict[str, Any] = Field(
        default_factory=dict, description="Parameters for the flow run."
    )
    idempotency_key: Optional[str] = Field(
        default=None,
        description=(
            "An optional idempotency key for the flow run. Used to ensure the same flow"
            " run is not created multiple times."
        ),
    )
    context: Dict[str, Any] = Field(
        default_factory=dict,
        description="Additional context for the flow run.",
        examples=[{"my_var": "my_val"}],
    )
    empirical_policy: FlowRunPolicy = Field(
        default_factory=FlowRunPolicy,
    )
    tags: List[str] = Field(
        default_factory=list,
        description="A list of tags on the flow run",
        examples=[["tag-1", "tag-2"]],
    )
    parent_task_run_id: Optional[UUID] = Field(
        default=None,
        description=(
            "If the flow run is a subflow, the id of the 'dummy' task in the parent"
            " flow used to track subflow state."
        ),
    )
    run_count: int = Field(
        default=0, description="The number of times the flow run was executed."
    )
    expected_start_time: Optional[DateTime] = Field(
        default=None,
        description="The flow run's expected start time.",
    )
    next_scheduled_start_time: Optional[DateTime] = Field(
        default=None,
        description="The next time the flow run is scheduled to start.",
    )
    start_time: Optional[DateTime] = Field(
        default=None, description="The actual start time."
    )
    end_time: Optional[DateTime] = Field(
        default=None, description="The actual end time."
    )
    total_run_time: datetime.timedelta = Field(
        default=datetime.timedelta(0),
        description=(
            "Total run time. If the flow run was executed multiple times, the time of"
            " each run will be summed."
        ),
    )
    estimated_run_time: datetime.timedelta = Field(
        default=datetime.timedelta(0),
        description="A real-time estimate of the total run time.",
    )
    estimated_start_time_delta: datetime.timedelta = Field(
        default=datetime.timedelta(0),
        description="The difference between actual and expected start time.",
    )
    auto_scheduled: bool = Field(
        default=False,
        description="Whether or not the flow run was automatically scheduled.",
    )
    infrastructure_document_id: Optional[UUID] = Field(
        default=None,
        description="The block document defining infrastructure to use this flow run.",
    )
    infrastructure_pid: Optional[str] = Field(
        default=None,
        description="The id of the flow run as returned by an infrastructure block.",
    )
    created_by: Optional[CreatedBy] = Field(
        default=None,
        description="Optional information about the creator of this flow run.",
    )
    work_queue_id: Optional[UUID] = Field(
        default=None, description="The id of the run's work pool queue."
    )

    work_pool_id: Optional[UUID] = Field(
        default=None, description="The work pool with which the queue is associated."
    )
    work_pool_name: Optional[str] = Field(
        default=None,
        description="The name of the flow run's work pool.",
        examples=["my-work-pool"],
    )
    state: Optional[State] = Field(
        default=None,
        description="The state of the flow run.",
        examples=["State(type=StateType.COMPLETED)"],
    )
    job_variables: Optional[dict] = Field(
        default=None,
        description="Job variables for the flow run.",
    )

    # These are server-side optimizations and should not be present on client models
    # TODO: Deprecate these fields

    state_type: Optional[StateType] = Field(
        default=None, description="The type of the current flow run state."
    )
    state_name: Optional[str] = Field(
        default=None, description="The name of the current flow run state."
    )

    def __eq__(self, other: Any) -> bool:
        """
        Check for "equality" to another flow run schema

        Estimates times are rolling and will always change with repeated queries for
        a flow run so we ignore them during equality checks.
        """
        if isinstance(other, FlowRun):
            exclude_fields = {"estimated_run_time", "estimated_start_time_delta"}
            return self.model_dump(exclude=exclude_fields) == other.model_dump(
                exclude=exclude_fields
            )
        return super().__eq__(other)

    @field_validator("name", mode="before")
    @classmethod
    def set_default_name(cls, name):
        return get_or_create_run_name(name)